### PR TITLE
8336846: assert(state->get_thread() == jt) failed: handshake unsafe conditions

### PR DESCRIPTION
**Pull request by sspitsyn**
The JVMTI Watch Field functions do not disable VTMS transitions with the `JvmtiVTMSTransitionDisabler`:
- `SetFieldAccessWatch()`
- `ClearFieldAccessWatch()`
- `SetFieldModificationWatch()`
- `ClearFieldModificationWatch()`
 so in the `recompute_enabled()` we could see that a vthread is mounted, but in the `EnterInterpOnlyModeClosure` handshake the thread could have been unmounted already. This is a root cause of failures with this assert.
 
The fix is to disable transitions in the `JvmtiEventControllerPrivate::change_field_watch()` function.

Testing:
- TBD: submit mach5 tiers 1-6

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336846](https://bugs.openjdk.org/browse/JDK-8336846): assert(state-&gt;get_thread() == jt) failed: handshake unsafe conditions (**Bug** - P3)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) 🔄 Re-review required (review applies to [bb94448c](https://git.openjdk.org/jdk/pull/20413/files/bb94448ca6730b10623a21bbbbeee27f9aa9a2cc))
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) 🔄 Re-review required (review applies to [3d166446](https://git.openjdk.org/jdk/pull/20413/files/3d1664463948acaec640e60d364f03736c80e1cf))
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) 🔄 Re-review required (review applies to [bb94448c](https://git.openjdk.org/jdk/pull/20413/files/bb94448ca6730b10623a21bbbbeee27f9aa9a2cc))
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20413/head:pull/20413` \
`$ git checkout pull/20413`

Update a local copy of the PR: \
`$ git checkout pull/20413` \
`$ git pull https://git.openjdk.org/jdk.git pull/20413/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20413`

View PR using the GUI difftool: \
`$ git pr show -t 20413`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20413.diff">https://git.openjdk.org/jdk/pull/20413.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20413#issuecomment-2261972014)

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiEventController.cpp [hunk 1]
```diff
@@ -985,12 +985,14 @@ JvmtiEventControllerPrivate::change_field_watch(jvmtiEvent event_type, bool adde
   if (added) {
     (*count_addr)++;
     if (*count_addr == 1) {
+      JvmtiVTMSTransitionDisabler disabler;
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiEventController.cpp:988]: Could you just declare this outside the if-block rather than repeating it?

Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiEventController.cpp:988]: Thank you for looking at this and for the suggestion, David.
I wanted to optimize a little bit but agree it is not worth it, so updated now.

**Review by plummercj — APPROVED** [review]
Looks good.

**Review by alexmenkov — APPROVED** [review]

**Review by pchilano — COMMENTED** [review]
Looks good, but I see we have ranking issues with JvmtiThreadState_lock now. We will have to change JvmtiVTMSTransition_lock to be safepoint-1.

**Review by lmesnik — APPROVED** [review]

**Comment by sspitsyn** [comment]
> Looks good, but I see we have ranking issues with JvmtiThreadState_lock now. We will have to change JvmtiVTMSTransition_lock to be safepoint-1.

Thank you, Patricio. Will check and test this.

**Comment by sspitsyn** [comment]
Chris, Alex and Leonid, thank you for review!

**Comment by pchilano** [comment]
> > Looks good, but I see we have ranking issues with JvmtiThreadState_lock now. We will have to change JvmtiVTMSTransition_lock to be safepoint-1.
> 
> Thank you, Patricio. Fixed as you suggested.
>
Thanks Serguei. You could also define it relative to JvmtiThreadState_lock with MUTEX_DEFL to make the dependency clear.

**Review by pchilano — APPROVED** [review]

**Comment by sspitsyn** [comment]
> Thanks Serguei. You could also define it relative to JvmtiThreadState_lock with MUTEX_DEFL to make the dependency clear.

Good suggestion, thanks. Thank you for review, Patricio.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks

**Comment by sspitsyn** [comment]
Thank you for review, David.

**Review by alexmenkov — APPROVED** [review]

**Pull request closed without merging**